jlm/llvm/opt/ScalarEvolution.hpp

```diff
@@ -0,0 +1,75 @@
+#ifndef JLM_LLVM_OPT_SCALAR_EVOLUTION_HPP
+#define JLM_LLVM_OPT_SCALAR_EVOLUTION_HPP
+
+#include <jlm/rvsdg/theta.hpp>
+#include <jlm/rvsdg/Transformation.hpp>
+
+#include <unordered_set>
+
+namespace jlm::llvm
+{
+class ScalarEvolution final : public jlm::rvsdg::Transformation
+{
+  class Statistics;
+
+public:
+  ~ScalarEvolution() noexcept override = default;
+
+  ScalarEvolution()
+      : Transformation("ScalarEvolution")
+  {}
+
+  ScalarEvolution(const ScalarEvolution &) = delete;
+
+  ScalarEvolution(ScalarEvolution &&) = delete;
+
+  ScalarEvolution &
+  operator=(const ScalarEvolution &) = delete;
+
+  ScalarEvolution &
+  operator=(ScalarEvolution &&) = delete;
+
+  void
+  Run(rvsdg::RvsdgModule & rvsdgModule, util::StatisticsCollector & statisticsCollector) override;
+
+  void
+  TraverseGraph(const rvsdg::Graph & rvsdg);
+
+  void
+  TraverseRegion(rvsdg::Region * region);
+
+  void
+  FindInductionVariables(const rvsdg::ThetaNode * thetaNode);
+
+private:
+  /**
+   * \brief Method used to check if the input to a node is a loop variable.
+   *
+   * \returns
+   *   An std::optional with the loop variable, if found.
+   */
+  static std::optional<rvsdg::ThetaNode::LoopVar>
+  TryGetLoopVarFromInput(
+      const rvsdg::Input * input,
+      std::vector<rvsdg::ThetaNode::LoopVar> loopVars)
+  {
+    for (auto & loopVar : loopVars)
+    {
+      if (input->origin() == loopVar.pre)
+      {
+        return loopVar;
+      }
+    }
+    return std::nullopt;
```

Comment:
I think this method can be implemented more efficiently. Have a look at [MapPreLoopVar()](https://github.com/phate/jlm/blob/5f36b59992b4f248584a35e40d94a6fa68528690/jlm/rvsdg/theta.cpp#L127C1-L135C2). It should be possible to use TryGetRegionParentNode() for this. I do not think there is a need to iterate through all loop variables.